### DISCUSSION TITLE
How to use BucketListener using ProxyManager

### DISCUSSION THREAD

**Discussion by naveen-maanju** [vladimir-bukhtoyarov/bucket4j#361] [Q&A] [answered]
I am trying to use the BucketListner using AbstractProxyManager, but didn't find a way through builder.  ANy help would be great!

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Currently you should use [toListenable](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/Bucket.java#L271) method.
```java
bucket = bucket.toListenable(listener);
```

**Reply by naveen-maanju** [reply]
@vladimir-bukhtoyarov , We are using the a customise version of CaffeneCacheManagerProxy as  we are using AT as key and want the  bucket TTL be same as AT life.

So do you mean to do something like this?
` 
BucketProxy bucketProxy = caffeneProxyManager.builder().build(keyId, bucketConfig);
bucketProxy.toListenable(listener);
`
If yes, will that not lead to adding a new listener every time, as it will create a new proxy instance instead of updating one in cache?

Correct me if I am wrong here.

**Comment by vladimir-bukhtoyarov** [comment]
like this:

```java
BucketProxy bucketProxy = caffeneProxyManager.builder().build(keyId, bucketConfig).toListenable(listener);
```

> will that not lead to adding a new listener every time, as it will create a new proxy instance instead of updating one in cache?

Listener is not persisted inside cache, Invocation of ```toListenable``` just creates new proxy.

**Comment by naveen-maanju** [comment]
Thanks a lot @vladimir-bukhtoyarov ! I will wait for builder approach where we can register listener while creating proxy from configuration and key.

**Reply by naveen-maanju** [reply]
Given I want to use listener per bucket instead of common for all bucket.

**Reply by vladimir-bukhtoyarov** [reply]
> Thanks a lot @vladimir-bukhtoyarov ! I will wait for builder approach where we can register listener while creating proxy from configuration and key.

It will be no difference with toListanable, so you should not wait anything.

**Reply by vladimir-bukhtoyarov** [reply]
> Given I want to use listener per bucket instead of common for all bucket.

You can achive it right now. Just use different listener for different bucket.

**Reply by naveen-maanju** [reply]
Yes, we can do that as well but, as we are using it in cloud gateway filter,  will that not lead to creation of a listener instance and two proxy instances for each request? Or it's not a costly operation?

**Reply by vladimir-bukhtoyarov** [reply]
Creation of proxy is cheap, this does not require an iteration with underlying storage(Caffeine in your case), and proxy does not hold something heavy.  According to listener, I do not know because it is managed by your code, if listener is heavyweight then it makes sense to cache listeners or redesign listeners to make it cheap.

**Reply by naveen-maanju** [reply]
Given we register the listener with MetricRegistry as well to capture the stats, will that not also require that we register the every new listener as well and eventually leading to large number of listeners and holding the reference of old(not in use listeners as well). In our use case I feel that listener be created  per bucket using configuration when we create the initial state of the bucket. But then how we will register them with MetricRegistry? Not sure, as you are the best person to decide here.

**Reply by vladimir-bukhtoyarov** [reply]
How does metric ID(and tags/labels if used) calculated for MetricsRegistry? What kind of Monitoring library is used?

**Reply by raghu298** [reply]
@naveen-maanju how you monitoring the bucket and what library you using?

**Reply by raghu298** [reply]
@vladimir-bukhtoyarov BucketListner takes only 1 parameter as token and how can we store metrics with userids?

**Reply by vladimir-bukhtoyarov** [reply]
Just create listener per bucket

**Reply by raghu298** [reply]
@vladimir-bukhtoyarov but how can i get other details to listner like userid. It has 1 parameter called token

**Reply by vladimir-bukhtoyarov** [reply]
It is your listener, pass additional context as you wish.

**Reply by raghu298** [reply]
@vladimir-bukhtoyarov what i checked to create a listener we need to implement BucketListener and override the methods and those are taking 1 paramter called token. Then how can we pass multi parmas.

Any sample example that can help?

Newbie to bucket4j sorry if asking silly questions

**Reply by vladimir-bukhtoyarov** [reply]
Use listener per context entry.

**Reply by raghu298** [reply]
@vladimir-bukhtoyarov any sample example to start with not able to get what is Listener per context entry?
I did not finding anything where to the listener call back methods can pass multi parameters apart from token.
what is listener per context?
